### PR TITLE
Improvements to bundling aggregations

### PR DESCRIPTION
**Pull request by clokep**
This somewhat came up [in a previous review](https://github.com/matrix-org/synapse/pull/11612#discussion_r780255443) and helps to avoid a potential issues similar to #11791. The main goal is to make the type hinting stronger and to stop mutating dictionaries in odd ways.

Should be reviewable commit-by-commit.

### CODE REVIEW

**Comment by clokep** [comment]
I believe this is failing lint for an unrelated reason.

**Review by DMRobertson — COMMENTED** [review]
Thanks, hope this wasn't too painful!

File: synapse/handlers/room.py
Comment (clokep) [reply synapse/handlers/room.py:1202]: It avoids creating a couple dictionaries to then `update()` but is mostly clean-up.

Once the changes for  #11825 land than this *is* a performance gain since fewer queries will occur.

Comment (clokep) [reply synapse/handlers/room.py:1202]: (It wasn't really related to the open PRs for #11825, so I put it in here since I was touching this code anyway, sorry about that!)

Comment (DMRobertson) [reply synapse/handlers/room.py:1202]: Thanks, that makes sense. And not to worry, I think it makes sense here.

Comment (DMRobertson) [reply synapse/handlers/room.py:1202]: Today's review is sponsored by the phrase "makes sense".

File: synapse/handlers/room.py
Comment (clokep) [reply synapse/handlers/room.py:1195]: Yeah, we are...I initially had converted that too and then stopped. I think it got messy, let me give it a try again though!

Comment (clokep) [reply synapse/handlers/room.py:1195]: I've updated this to not mutate those results, it got a bit messy though. Let me know if you'd prefer I back it out. See 177783111825eb78c76c797fe033dd0ee1b2311d.

File: synapse/handlers/search.py
Comment (clokep) [inline synapse/handlers/search.py:429]: Yes, this still mutates the dictionary, but it seems rather messy to *not* do this...

File: tests/rest/client/test_relations.py
Comment (clokep) [inline tests/rest/client/test_relations.py:580]: This was missed in #11791, oops.

File: synapse/storage/databases/main/relations.py
Comment (clokep) [inline synapse/storage/databases/main/relations.py:694]: This change means that if there are no bundled aggregations we don't include it at all, instead of including an empty object.

This was always the intention and what all the docstrings say, but guess the logic was slightly wrong. I noticed this when I was debugging something else.

**Comment by DMRobertson** [comment]
I think this needs develop merging in again :(

**Comment by clokep** [comment]
> I think this needs develop merging in again :(

Done!

File: synapse/handlers/room.py
Comment (DMRobertson) [reply synapse/handlers/room.py:1195]: To be honest I don't mind too much either way. (I was mostly checking that we hadn't missed a bit of work that we'd intended to do, judging by the commit message).

I like that the change makes it a bit easier to see what you get out of `get_events_around` (i.e. a thing with specific fields rather than an arbitrary dictionary), but I don't really have a strong opinion here. Happy to defer to your judgement!

**Review by DMRobertson — APPROVED** [review]
LGTM!

As per https://github.com/matrix-org/synapse/pull/11815#discussion_r792612107, I don't mind if we commit this with or without 177783111825eb78c76c797fe033dd0ee1b2311d. I'll leave that one to your discretion.

**Pull request merged**
